docs/source/data/fetching-data.mdx

```diff
@@ -1,6 +1,217 @@
+
+## Creating data source classes
+
+Your data source class can be as straightforward or complex as you need it to be. You know what data your server needs, and you can let that be the guide for the methods your class should include. 
+
+Below is an example of a data source class that connects to a database:
+
+<MultiCodeBlock>
+
+```ts title="bookings.ts"
+export class BookingsDataSource {
+  private dbConnection;
+  private context;
+
+  constructor(options: { context: { token: string } }) {
+    this.dbConnection = this.initializeDBConnection();
+    this.context = options.context;
+  }
+
+  async initializeDBConnection() {
+    // set up our database details, instantiate our connection,
+    // and return that database connection
+    return dbConnection;
+  }
+
+  async getBooking(bookingId) {
+    return await this.db.Booking.findByPk(bookingId);
+  }
+
+  //... more methods for finding and creating bookings
+}
+```
+
+</MultiCodeBlock>
+
+> Apollo's [`RestDataSource`](https://github.com/apollographql/datasource-rest/blob/main/src/RESTDataSource.ts) class is a built-out example of how a data source class can handle caching, deduplication, and errors.
+
+## Adding data sources to your context
+
+You can add data sources to your server's `context` initialization function, like so:
+
+<MultiCodeBlock>
+
+```ts title="index.ts"
+//highlight-start
+interface ContextValue {
+  dataSources: {
+    dogsDB: DogsDataSource;
+    catsApi: CatsAPI;
+  };
+  token: string
+}
+//highlight-end
+
+const server = new ApolloServer<ContextValue>({
+  typeDefs,
+  resolvers,
+});
+
+const { url } = await startStandaloneServer(server, {
+  context: async () => {
+    const { cache } = server; // highlight-line
+    const token = req.headers.token;
+    return {
+      // We create new instances of our data sources with each request.
+      // We can pass in our server's cache, context, or any other
+      // info our data sources require. 
+      // highlight-start
+      dataSources: {
+        dogsDB: new DogsDataSource({ cache, context: token }),
+        catsApi: new CatsAPI({ cache }),
+      },
+      //highlight-end
+      token
+    };
+  },
+});
+
+console.log(`🚀  Server ready at ${url}`);
+```
+
+</MultiCodeBlock>
+
+<!-- TODO(AS4) add link to context article once exists  -->
+Apollo Server calls [the `context` initialization](./resolvers/#the-context-argument) function for _every incoming operation_. This means:
+- For every operation, `context` returns an _object_ containing new instances of your data source classes (in this case, `DogsDataSource` and `CatsAPI`).
+- The **`context` function should create a new instance of each data source class for each operation.** You might accidentally combine results from multiple operations if multiple operations share a single data source instance.
+
+Your resolvers can then access your data sources from the shared `context` object and use them to fetch data:
```

Comment:
```suggestion
Your resolvers can then access your data sources from the shared `contextValue` object and use them to fetch data:
```

Reply:
Resolved this in my subsequent commit :)